vembrane/modules/filter.py

```diff
@@ -160,67 +166,127 @@ def filter_vcf(
+            is_bnd = has_svtype and record.info.get("SVTYPE", None) == "BND"
+            if is_bnd:
+                mate_ids = record.info.get("MATEID", [])
+                event_name = record.info.get("EVENT", None)
+
+                if len(mate_ids) > 1 and not event_name:
+                    raise ValueError(
+                        f"Filtering of BND records with multiple mates is unsupported "
+                        f"(see VCF 4.3, section 5.4.3 'Multiple mates'):\n{str(record)}"
+                    )
+
+                # if we have a mate pair, use their sorted ids as a key
+                mate_id = mate_ids[0] if len(mate_ids) == 1 else None
+                mate_pair = mate_key([record.id, mate_id]) if mate_ids else None
+
+                # if EVENT is set, it has priority over MATEID.
+                event_name = event_name or mate_pair
+
+                # if both EVENT and MATEID are not set
+                # we can only treat this BND record as a regular one,
+                # (or somehow try to guess its correct mate)
+                if not event_name:
+                    print(
+                        f"Warning: Encountered breakend record at index {idx} "
+                        f"without either of MATEID or EVENT specified, "
+                        f"treating it as a regular record:\n{str(record)}",
+                        file=stderr,
+                    )
+                    if record_has_passed:
                         yield record
+                    continue
+
+                event = events.get(event_name, None)
+
+                # if there's already an associated event
+                if event:
+                    # add this record to the event
+                    event.add(record, record_has_passed)
+
+                    # if we already know that the event is a "PASS"…
+                    if event.passed:
+                        # … emit all records associated with it
+                        yield from event.emit()
+
+                        # in the case of a simple mate pair, we can delete the event
+                        # at this point, because no more records will be added to it
+                        if event.is_mate_pair():
+                            del events[mate_pair]
+                else:
+                    # if there's no entry for the event or mate pair yet, create one
+                    is_mate_pair = mate_pair and mate_pair == event_name
+                    event = BreakendEvent(event_name, is_mate_pair)
+                    event.add(record, record_has_passed)
+                    events[event_name] = event
+            elif record_has_passed:
+                yield record
+
+        if len(events) > 0:
+            # output BNDs if any are were left unprocessed
+            for event_name, event in events.items():
+                if event_name and event.passed:
+                    yield from event.emit()
     else:
```

Comment:
Just for readability:

```suggestion
    # this else implies `preserve_order` is set
    else:
```

Or make it explicit in the code:
```suggestion
    elif preserve_order:
```